File: tests/playground/expr_const_folding.ml
type expr =
| Var of int
| Val of int
| Add of expr * expr
| Mul of expr * expr;;

let dec n =
  if n == 0 then 0 else n - 1;;

let rec mk_expr n v =
  if n == 0 then (if v == 0 then Var 1 else Val v)
  else Add (mk_expr (n-1) (v+1), mk_expr (n-1) (dec v));;

let rec append_add e1 e2 =
match (e1, e2) with
| (Add (e1, e2), e3) -> Add (e1, append_add e2 e3)
| (e1, e2)           -> Add (e1, e2);;

let rec append_mul e1 e2 =
match (e1, e2) with
| (Mul (e1, e2), e3) -> Mul (e1, append_mul e2 e3)
| (e1, e2)           -> Mul (e1, e2);;

let rec reassoc e =
match e with
| Add (e1, e2) ->
  let e1' = reassoc e1 in
  let e2' = reassoc e2 in
  append_add e1' e2'
| Mul (e1, e2) ->
  let e1' = reassoc e1 in
  let e2' = reassoc e2 in
  append_mul e1' e2'
| e -> e;;

let rec const_folding e =
match e with
| Add (e1, e2) ->
  let e1 = const_folding e1 in
  let e2 = const_folding e2 in
  (match (e1, e2) with
   | (Val a, Val b)          -> Val (a+b)
   | (Val a, Add (e, Val b)) -> Add (Val (a+b), e)
   | (Val a, Add (Val b, e)) -> Add (Val (a+b), e)
   | _                       -> Add (e1, e2))
| Mul (e1, e2) ->
  let e1 = const_folding e1 in
  let e2 = const_folding e2 in
  (match (e1, e2) with
   | (Val a, Val b)          -> Val (a*b)
   | (Val a, Mul (e, Val b)) -> Mul (Val (a*b), e)
   | (Val a, Mul (Val b, e)) -> Mul (Val (a*b), e)
   | _                       -> Mul (e1, e2))
| e ->  e;;

let rec size e =
match e with
| Add (e1, e2) -> size e1 + size e2 + 1
| Mul (e1, e2) -> size e1 + size e2 + 1
| e            -> 1;;

let rec eeval e =
 match e with
 | Val n -> n
 | Var x -> 0
 | Add (e1, e2) -> eeval e1 + eeval e2
 | Mul (e1, e2) -> eeval e1 * eeval e2;;

let e  = (mk_expr 23 1) in
let v1 = eeval e in
let v2 = eeval (const_folding (reassoc e)) in
Printf.printf "%8d %8d\n" v1 v2;;

File: tests/playground/expr_const_folding.hs
data Expr = Var Integer
  | Val Integer
  | Add Expr Expr
  | Mul Expr Expr

mk_expr :: Integer -> Integer -> Expr
mk_expr 0 v = if v == 0 then Var 1 else Val v
mk_expr n v = Add (mk_expr (n-1) (v+1)) (mk_expr (n-1) (max (v-1) 0))

append_add :: Expr -> Expr -> Expr
append_add (Add e₁ e₂) e₃ = Add e₁ (append_add e₂ e₃)
append_add e₁          e₂ = Add e₁ e₂

append_mul :: Expr -> Expr -> Expr
append_mul (Mul e₁ e₂) e₃  = Mul e₁ (append_mul e₂ e₃)
append_mul e₁          e₂ = Mul e₁ e₂

reassoc :: Expr -> Expr
reassoc (Add e₁ e₂) =
  let e₁' = reassoc e₁ in
  let e₂' = reassoc e₂ in
  append_add e₁' e₂'
reassoc (Mul e₁ e₂) =
  let e₁' = reassoc e₁ in
  let e₂' = reassoc e₂ in
  append_mul e₁' e₂'
reassoc e = e

const_folding :: Expr -> Expr
const_folding (Add e₁ e₂) =
  let e₁' = const_folding e₁ in
  let e₂' = const_folding e₂ in
  (case (e₁', e₂') of
     (Val a, Val b        ) -> Val (a+b)
     (Val a, Add e (Val b)) -> Add (Val (a+b)) e
     (Val a, Add (Val b) e) -> Add (Val (a+b)) e
     (_,     _            ) -> Add e₁' e₂')
const_folding (Mul e₁ e₂) =
  let e₁' = const_folding e₁ in
  let e₂' = const_folding e₂ in
  (case (e₁', e₂') of
     (Val a, Val b        ) -> Val (a*b)
     (Val a, Mul e (Val b)) -> Mul (Val (a*b)) e
     (Val a, Mul (Val b) e) -> Mul (Val (a*b)) e
     (_,     _            ) -> Mul e₁' e₂')
const_folding e         = e

eval :: Expr -> Integer
eval (Var _) = 0
eval (Val v) = v
eval (Add l r) = eval l + eval r
eval (Mul l r) = eval l * eval r

main :: IO ()
main =
  let e  = (mk_expr 23 1) in
  let v₁ = eval e in
  let v₂ = eval (const_folding (reassoc e)) in
  putStrLn (show v₁ ++ " " ++ show v₂)
